src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp

```diff
@@ -0,0 +1,443 @@
+                                                 });
+    if (lookupResult != CHIP_ERROR_NOT_FOUND)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::AlreadyExists);
+        return;
+    }
+
+    ProvisionRootCertificateResponse::Type response;
+    auto status = mDelegate.ProvisionRootCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req,
+                                              response.caid);
+
+    if (!status.IsSuccess())
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    VerifyOrDieWithMsg(response.caid <= kMaxRootCertId, NotSpecified, "Spec requires CAID to be < kMaxRootCertId");
+    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+}
+
+void TlsCertificateManagementServer::HandleFindRootCertificate(HandlerContext & ctx, const FindRootCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: FindRootCertificate");
+    CHIP_ERROR result;
+    if (req.caid.IsNull())
+    {
+        result = mDelegate.RootCertsForFabric(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(),
+                                              [&](auto & certs) -> CHIP_ERROR {
+                                                  FindRootCertificateResponse::Type response;
+                                                  response.certificateDetails = certs;
+                                                  ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+                                                  return CHIP_NO_ERROR;
+                                              });
+    }
+    else
+    {
+        result = mDelegate.FindRootCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(),
+                                        req.caid.Value(), [&](auto & certificate) -> CHIP_ERROR {
+                                            FindRootCertificateResponse::Type response;
+                                            DataModel::List<const TLSCertStruct::Type> details(&certificate, 1);
+                                            response.certificateDetails = details;
+                                            ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+                                            return CHIP_NO_ERROR;
+                                        });
+    }
+    if (result == CHIP_ERROR_NOT_FOUND)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+    }
+    else if (result != CHIP_NO_ERROR)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+    }
+}
+
+void TlsCertificateManagementServer::HandleLookupRootCertificate(HandlerContext & ctx,
+                                                                 const LookupRootCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: LookupRootCertificate");
```

Comment:
Where is the spec-required constraint check on req.fingerprint?

Reply:
Apologies, this is added @ https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/11899 (we found it while reviewing this PR)